### PR TITLE
Add "RHCOS Ignition Fail to Live"

### PR DESCRIPTION
**Pull request by darkmuggle**
Proposal to change RHCOS to support a "fail to a live root", to drop UPI installed platforms into an interactive recovery console.

### CODE REVIEW

File: enhancements/rhcos/rhcos-fail-live.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+* Discard all state upon leaving the environmet
+* Provide disk-based UPI installation on unsupported platforms using the "metal" target.
+
+### Non-goals
+
+All other tasks.
+
+## Proposal
+
+Rather than failing to the `emergency.target` upon an Ignition Failure.
```
Comment (ashcrow) [inline enhancements/rhcos/rhcos-fail-live.md]: Remove `.`

File: enhancements/rhcos/rhcos-fail-live.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+Specific platforms (Qemu and Metal) will have platform-specific configurations
```
Comment (ashcrow) [inline enhancements/rhcos/rhcos-fail-live.md]: I think this is continuing from previous fragment.

**Comment by miabbott** [comment]
Notes from an internal discussion around the proposal - https://hackmd.io/JUzhpTJaRtCD0LdJNaOsHQ

**Comment by cgwalters** [comment]
> Notes from an internal discussion around the proposal - https://hackmd.io/JUzhpTJaRtCD0LdJNaOsHQ

Awesome, thanks a ton for taking these notes and posting them.  I think it's critically important that we continually fight the tendency to discuss and make decisions behind the RHT firewall and instead behave as much as possible as part of a collaborative FOSS project.

A good perspective on this is [this blog post about Rust and the "core team"](http://aturon.github.io/tech/2018/05/25/listening-part-1/).

In particular:

> No New Rationale: decisions must be made only on the basis of rationale already debated in public (to a steady state)

So if we didn't post about it publicly, we shouldn't be making any decisions based on it.

**Comment by darkmuggle** [comment]
We don't have consensus yet. Another model was proposed that may supersede this idea. 
/hold

File: enhancements/rhcos/rhcos-fail-live.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+*Adding error handlers to Ignition*: this was rejected as it would require a
+significant re-architecture. Ignition was designed to treat all errors as fatal.
+
+*Attempting to discern the fault causing the error* (such as no networking) and emitting
+a totem. This idea would allow an external actor to change how things boot.
+The idea is similar to the proposed solution but it would require more logic in Dracut.
+
+*Using the CoreOS Installer to drive network information* was considered and rejected
+since it is specific to "metal" images. With this solution, its concievable to use
+UPI installations on unsupported platforms such as Azure-like installations.
```
Comment (miabbott) [inline enhancements/rhcos/rhcos-fail-live.md]: `s/Azure/Hyper-V/`

**Comment by cgwalters** [comment]
OK so Dusty's suggestion is basically "Just use the Live ISO" even for e.g. the VMWare case which is cool because we're already [on track to ship the Live ISO](https://github.com/openshift/enhancements/pull/210) so there'd be nothing new at all to ship for "give me something to boot to a console where I can generate a network and Ignition config".

The only detail though is we need to [handle networking in the no-config case](https://github.com/coreos/ignition/pull/948).

**Comment by darkmuggle** [comment]
I'm not familiar with enough with the installer ISO to flesh out @dustymabe's idea. And I'm not sure the notes captured the nuance of it.

**Comment by miabbott** [comment]
I think we can provide this "fail to live" functionality as well as providing the Live ISO to do similar kinds of configuration.  One does not appear to preclude the other.

I'd recommend we continue to pursue this proposal, as we have received early feedback from folks in the field that it is desirable.

**Comment by cgwalters** [comment]
It's really simple, for vmware or bare metal:

- Download pristine `-live.iso` 
- Boot from it (image to USB, add as device in vmware, add via IMPI on metal, etc.)
- You get auto-logged into a live CoreOS system without network
- Generate an Ignition config however you like, plus e.g. kernel cmdline arguments for network
- Run `coreos-installer install -i /path/to/config.ign --firstboot-args <network kargs> /dev/whatever`

All of this exists today and works except for the last issue mentioned in https://github.com/openshift/enhancements/pull/256#issuecomment-603491533

What feedback from the field isn't covered by this?

**Comment by cgwalters** [comment]
That said, linking this to the "nmtui" thing - we should ship that as part of the live ISO, and then have support for synthesizing the network config into kernel args - *or* special support for dropping network config into the initramfs.

**Comment by darkmuggle** [comment]
In follow up discussions, this idea has been superseded for an installer-based path that would handle injecting setting the network configuration.

**Comment by miabbott** [comment]
> It's really simple, for vmware or bare metal:
> 
>     * Download pristine `-live.iso`
> 
>     * Boot from it (image to USB, add as device in vmware, add via IMPI on metal, etc.)
> 
>     * You get auto-logged into a live CoreOS system without network
> 
>     * Generate an Ignition config however you like, plus e.g. kernel cmdline arguments for network
> 
>     * Run `coreos-installer install -i /path/to/config.ign --firstboot-args <network kargs> /dev/whatever`
> 
> 
> All of this exists today and works except for the last issue mentioned in [#256 (comment)](https://github.com/openshift/enhancements/pull/256#issuecomment-603491533)

Do we need a separate enhancement to cover the behavior change of dropping to a live system?

**Comment by cgwalters** [comment]
> Do we need a separate enhancement to cover the behavior change of dropping to a live system?

That's just it - there isn't a behavior change, it's what the Live ISO you can download from e.g. https://getfedora.org/coreos/ does *today* (and will for RHCOS too).

**Comment by miabbott** [comment]
> That's just it - there isn't a behavior change, it's what the Live ISO you can download from e.g. https://getfedora.org/coreos/ does _today_ (and will for RHCOS too).

So any enhancement required would really be #210

**Comment by darkmuggle** [comment]
Per https://meetbot.fedoraproject.org/fedora-meeting-1/2020-03-25/fedora_coreos_meeting.2020-03-25-16.31.html
This idea is done.

**Comment by cgwalters** [comment]
> So any enhancement required would really be #210

Well...the *basics* are that yes.  It avoids the "catch the grub prompt" required [in this blog](https://blog.openshift.com/openshift-4-2-vsphere-install-with-static-ips/).

We could call it done there.  But I think there's a *lot* more we could do beyond solving the grub prompt problem, including more easily generating those kargs, and dealing with fetching the base Ignition configs.

And all that stuff could be a separate enhancement or we could append it to the existing one.

**Pull request closed without merging**